### PR TITLE
TOMAS updates: Capability to run in GCHP, NetCDF diagnostics, and other fixes

### PR DESCRIPTION
**Pull request by msulprizio**
### Name and Institution (Required)

Name: Melissa Sulprizio
Institution: Harvard / GCST

### Describe the update

This pull request combines many updates and fixes for the TOMAS simulation that were originally implemented by @BettyCroft including:
- https://github.com/geoschem/geos-chem/pull/1378
  - includes implementation of TOMAS netCDF diagnostics
- https://github.com/geoschem/geos-chem/pull/1569
- https://github.com/geoschem/geos-chem/pull/1959
- https://github.com/geoschem/geos-chem/pull/1960

This pull request should be merged alongside:
- https://github.com/geoschem/HEMCO/pull/251
- https://github.com/geoschem/GCHP/pull/359

BPCH diagnostics can now be retired following this update. A separate pull request will be submitted with those updates.

### Expected changes

This pull request will cause small differences in the benchmark simulation due to the addition of dummy species in the full-chemistry mechanism as discussed in https://github.com/geoschem/geos-chem/issues/1742.

### CODE REVIEW

File: GeosCore/tomas_mod.F90 [hunk 1]
```diff
@@ -1745,19 +1638,28 @@ SUBROUTINE getCondSink(Nko, Mko, spec, CS, sinkfrac, surf_area, &
...
        surf_area = surf_area+Nko(k)*pi*(Dpk(k)*1.0e+6_fp)**2
     enddo
+    !bc 21/01/2022 - check if divide by zero below -added 2 if 
     do k=1,ibins
-       sinkfrac(k) = Dpk(k)*Nko(k)*beta(k)/CS
+       if (CS > 0.e-0_fp) then
+          sinkfrac(k) = Dpk(k)*Nko(k)*beta(k)/CS
+       else
+          sinkfrac(k) = 0.e-0_fp
+       endif
```
Comment (yantosca) [inline GeosCore/tomas_mod.F90:1655]: We can get rid of the ELSE with:
```F90
    do k=1,ibins
       sinkfrac(k) = 0.0_fp
       if (CS > 0.0_fp) then
          sinkfrac(k) = Dpk(k)*Nko(k)*beta(k)/CS
       endif
    enddo
```

File: GeosCore/tomas_mod.F90 [hunk 2]
```diff
@@ -1745,19 +1638,28 @@ SUBROUTINE getCondSink(Nko, Mko, spec, CS, sinkfrac, surf_area, &
...
     enddo
-    CS = 2.e+0_fp*pi*dble(Di)*CS/(dble(boxvol)*1e-6_fp)
-    surf_area = surf_area/(dble(boxvol))
+    CS = 2.e+0_fp*pi*dble(Di)*CS/(dble(boxvol)*1.e-6_fp)
+    if (CS  > 0.e-0_fp) then
+       surf_area = surf_area/(dble(boxvol))
+    else
+       surf_area = 0.e-0_fp
+    endif
```
Comment (yantosca) [inline GeosCore/tomas_mod.F90:1662]: We can get rid of the ELSE with:
```F90
    surf_area = 0.0_fp
    if (CS  > 0.0_fp) then
       surf_area = surf_area/(dble(boxvol))
    endif
```

File: GeosCore/tomas_mod.F90 [hunk 3]
```diff
@@ -4051,130 +3965,160 @@ SUBROUTINE SOACOND( MSOA, I, J, L, BOXVOL, TEMPTMS, PRES, &
...
+       ! the negative value (win, 9/12/05)
+       !-----------------------------------------------------------------------
+
+       IF( PDBG ) THEN
+          !print 12, I,J,L
+12        FORMAT( 'Error in SOAcond at ', 3I4 )
+          if( .not. SPINUP(60.) )write(116,*) 'Error in SOACOND at',i,j,l
        ELSE
-          OCTAU(K) = 0.e+0_fp
+          PDBG = .false.
```
Comment (yantosca) [inline GeosCore/tomas_mod.F90:4072]: I think the ELSE is here superfluous. PDBG is already FALSE to get to the ELSE block.

File: Headers/state_diag_mod.F90 [hunk 4]
```diff
@@ -16383,6 +18094,14 @@ SUBROUTINE Get_Mapping( Input_Opt,   State_Chm, TaggedDiagList,            &
              READ( TagItem%Name(S:S), '(I1)' ) index
              mapData%slot2id(TagItem%index) = index
 
+          ELSEIF ( isTomasBin ) THEN
+
+             ! TomasSbin: Tag names are "bin1" .. "bin15", so the
+             ! bin number is the last 2 characters of the tag name
```
Comment (yantosca) [inline Headers/state_diag_mod.F90:18100]: Update comment: `"bin01" .. "bin15"`

**Review by yantosca — CHANGES_REQUESTED** [review]
This PR mostly involves:
- Changing `IBINS` to `State_Chm%nTomasBins`
- Changing `NK1, NK2, ...`, to `NK01, NK02, ...`
- Adding History diagnostics for TOMAS

This PR is good to merge after fixing a few issues.  Also, we would need to update `KPP/fullchem/CHANGELOG_fullchem.md` to note that the mechanism has been changed by adding PH2SO4 into the rxns.

File: GeosCore/tomas_mod.F90 [hunk 5]
```diff
@@ -672,9 +569,9 @@ SUBROUTINE AEROPHYS( Input_Opt, State_Chm, State_Grid, State_Met, RC )
 
        ! Swap Spc into Nk, Mk, Gc arrays
        DO N = 1, IBINS
-          NK(N) = Spc(id_NK1-1+N)%Conc(I,J,L)
+          NK(N) = Spc(id_NK01-1+N)%Conc(I,J,L)
```
[the remark this answers is not in the document]
Comment (msulprizio) [reply GeosCore/tomas_mod.F90:572]: Yes, in `INIT_TOMAS` the saved module variable IBINS is set to:

```
    ! Number of size bins
    IBINS = State_Chm%nTomasBins
```

File: GeosCore/fullchem_mod.F90 [hunk 6]
```diff
@@ -2628,6 +2778,12 @@ SUBROUTINE Init_FullChem( Input_Opt, State_Chm, State_Diag, RC )
     id_SALAAL   = Ind_( 'SALAAL'       )
     id_SALC     = Ind_( 'SALC'         )
     id_SALCAL   = Ind_( 'SALCAL'       )
+#ifdef TOMAS
+    id_NK5      = Ind_( 'NK5'          )
```
[the remark this answers is not in the document]
Comment (msulprizio) [reply GeosCore/fullchem_mod.F90]: Now fixed in https://github.com/geoschem/geos-chem/pull/2060/commits/2d6ee6630102a6a11b7643d9d4094a12c9fa453c.

File: GeosCore/tomas_mod.F90 (hunk 1 above)
Comment (msulprizio) [reply GeosCore/tomas_mod.F90:1654]: This change bas been made in https://github.com/geoschem/geos-chem/pull/2060/commits/2d6ee6630102a6a11b7643d9d4094a12c9fa453c.

File: GeosCore/tomas_mod.F90 (hunk 2 above)
Comment (msulprizio) [reply GeosCore/tomas_mod.F90:1660]: This change bas been made in https://github.com/geoschem/geos-chem/pull/2060/commits/2d6ee6630102a6a11b7643d9d4094a12c9fa453c.

**Review by yantosca — APPROVED** [review]
Thanks @msulprizio!  Good to merge.

**Pull request merged**